text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,395 @@
+
+### The common data types
+
+- Key: raw binary data
+- Value: raw binary data
+- KvPair: Key-value pair type
+- KeyRange: Half-open interval of keys
+- Config: Configuration for client
+
+### Raw Key-Value API Basic Usage
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+    ```rust
+        let config = Config::new(vec!["127.0.0.1:2379"]);
+    ```
+
+2. Create a Raw Key-Value client.
+
+    ```rust
+        let client = RawClient::new(&config);
+    ```
+
+3. Call the Raw Key-Value client methods to access the data on TiKV. The Raw Key-Value API contains following methods
+
+    ```rust
+        fn get<K, C>(&self, key: K, cf: C) -> KvFuture<Value>
+        where
+            K: AsRef<Key>,
+            C: Into<Option<String>>;
+
+        fn batch_get<K, C>(&self, keys: K, cf: C) -> KvFuture<Vec<KvPair>>
+        where
+            K: AsRef<[Key]>,
+            C: Into<Option<String>>;
+
+        fn put<P, C>(&self, pair: P, cf: C) -> KvFuture<()>
+        where
+            P: Into<KvPair>,
+            C: Into<Option<String>>;
+
+        fn batch_put<I, P, C>(&self, pairs: I, cf: C) -> KvFuture<()>
+        where
+            I: IntoIterator<Item = P>,
+            P: Into<KvPair>,
+            C: Into<Option<String>>;
+
+        fn delete<K, C>(&self, key: K, cf: C) -> KvFuture<()>
+        where
+            K: AsRef<Key>,
+            C: Into<Option<String>>;
+
+        fn batch_delete<K, C>(&self, keys: K, cf: C) -> KvFuture<()>
+        where
+            K: AsRef<[Key]>,
+            C: Into<Option<String>>;
+
+        fn scan<R, C>(&self, range: R, limit: u32, key_only: bool, cf: C) -> KvFuture<Vec<KvPair>>
+        where
+            R: RangeBounds<Key>,
+            C: Into<Option<String>>;
+
+        fn batch_scan<R, B, C>(
+            &self,
+            ranges: R,
+            each_limit: u32,
+            key_only: bool,
+            cf: C,
+        ) -> KvFuture<Vec<KvPair>>
+        where
+            R: AsRef<[B]>,
+            B: RangeBounds<Key>,
+            C: Into<Option<String>>;
+
+        fn delete_range<R, C>(&self, range: R, cf: C) -> KvFuture<()>
+        where
+            R: RangeBounds<Key>,
+            C: Into<Option<String>>;
+    ```
+
+#### Usage example of the Raw Key-Value API
+
+```rust
+    extern crate futures;
+    extern crate tikv_client;
+
+    use futures::future::Future;
+    use tikv_client::raw::Client;
+    use tikv_client::*;
+
+    fn main() {
+        let config = Config::new(vec!["127.0.0.1:3379"]);
+        let raw = raw::RawClient::new(&config)
+            .wait()
+            .expect("Could not connect to tikv");
+
+        let key: Key = b"Company".to_vec().into();
+        let value: Value = b"PingCAP".to_vec().into();
+
+        raw.put((Clone::clone(&key), Clone::clone(&value)), None)
```

Comment:
Adding `None` to all the API calls is unsatisfying. Does it make sense to put the column family in the `Config`? Alternatively, if we have such a small core API surface we could expose two functions with the different signatures.

Reply:
Hm, maybe adding another method sounds better? User may use multiple column family for different data, so it makes better sense to let user choose which cf to use at per call basis.

Reply:
Like put_with_cf or put_with_column_family

Reply:
I would probably favor `put_cf` so the method name is not really long. Or if we should encourage use of column families, make that the shorter api and have `put_def` for using the default column family.

I would be interested in seeing the builder style. I am actually wondering if it makes sense to try to build it into the Key type since it partitions keys.

Reply:
CF is only available in Raw KV interface. So we can move Key/Value/KvPair definition to raw and transactional module so they can have different definition.

Reply:
I'm planning to change it to request builder style interface. Haven't done it yet, will try to fix couple of issues we discussed all together.

Reply:
None is eliminated by builder api